compiler/rustc_passes/src/check_attr.rs

```diff
@@ -684,9 +684,30 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
     }
 
     /// Checks if the `#[non_exhaustive]` attribute on an `item` is valid.
-    fn check_non_exhaustive(&self, hir_id: HirId, attr: &Attribute, span: Span, target: Target) {
+    fn check_non_exhaustive(
+        &self,
+        hir_id: HirId,
+        attr: &Attribute,
+        span: Span,
+        target: Target,
+        item: Option<ItemLike<'_>>,
+    ) {
         match target {
-            Target::Struct | Target::Enum | Target::Variant => {}
+            Target::Struct => {
+                if let Some(ItemLike::Item(hir::Item {
+                    kind: hir::ItemKind::Struct(hir::VariantData::Struct { fields, .. }, _),
+                    ..
+                })) = item
+                    && !fields.is_empty()
+                    && fields.iter().any(|f| f.default.is_some())
+                {
+                    self.dcx().emit_err(errors::NonExhaustiveWithDefaultFieldValues {
+                        attr_span: attr.span,
+                        defn_span: span,
+                    });
```

Comment:
Question [GATE 1/2]: does this check need to be feature-gated? As in, if I write

```rs
#[derive(Default)]
#[non_exhaustive]
struct Foo {
    x: i32 = 42 + 3,
}
```

This error is still emitted even though I did not enable the feature. It's not a *significant* problem because default field values are already feature-gated at parsing time.

Reply:
Yeah. My rationale was that gating this error was redundant, as it is not something we want to allow to begin with (and might change behavior before stabilization of the feature). The biggest benefit of gating it would be to reduce the number of errors (1 instead of 2), and that didn't seem like a good thing necessarily (telling users "yeah, enable the feature" followed by "lol, no, you can't do that"). What do you think?

FWIW, the current output:

```
error[E0658]: default values on fields are experimental
 --> foo.rs:4:11
  |
4 |     x: i32 = 42 + 3,
  |           ^^^^^^^^^
  |
  = note: see issue #132162 <https://github.com/rust-lang/rust/issues/132162> for more information
  = help: add `#![feature(default_field_values)]` to the crate attributes to enable

error: `#[non_exhaustive]` can't be used to annotate items with default field values
 --> foo.rs:2:1
  |
2 |   #[non_exhaustive]
  |   ^^^^^^^^^^^^^^^^^
3 | / struct Foo {
4 | |     x: i32 = 42 + 3,
5 | | }
  | |_- this struct has default field values
```